examples/cpp/customserializable/main.cpp

```diff
@@ -0,0 +1,133 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include <sstream>
+
+#include <geode/CacheFactory.hpp>
+#include <geode/PoolManager.hpp>
+#include <geode/PdxWrapper.hpp>
+
+using namespace apache::geode::client;
+
+class Order : public PdxSerializable {
+ public:
+  Order() {
+    orderID = (unsigned long)0;
+    quantity = (unsigned int)0;
+    name = "";
+  };
+
+  Order(unsigned long inOrderID, const std::string& inName,
+        unsigned int inQuantity)
+      : orderID(inOrderID), name(inName), quantity(inQuantity){};
+
+  virtual ~Order(){};
+
+  unsigned long getOrderID() { return orderID; }
+  const std::string& getName() const { return name; }
+  unsigned int getQuantity() { return quantity; }
+
+  virtual void fromData(PdxReader& pdxReader) override {
+    try {
+      orderID = (unsigned long)pdxReader.readLong(orderid_key);
+      name = pdxReader.readString(name_key);
+      quantity = (unsigned int)pdxReader.readInt(quantity_key);
+    } catch (std::exception& e) {
+      std::cout << "Caught exception: " << e.what() << std::endl;
+      return;
+    }
+  }
+
+  virtual void toData(PdxWriter& pdxWriter) const override {
+    try {
+      pdxWriter.writeLong(orderid_key, orderID);
+      pdxWriter.markIdentityField(orderid_key);
+
+      pdxWriter.writeString(name_key, name);
+      pdxWriter.markIdentityField(name_key);
+
+      pdxWriter.writeInt(quantity_key, quantity);
+      pdxWriter.markIdentityField(quantity_key);
+
+    } catch (std::exception& e) {
+      std::cout << "Caught exception: " << e.what() << std::endl;
+      return;
+    }
+    return;
+  }
+
+  static PdxSerializable* createDeserializable() { return new Order(); }
+
+  virtual std::string toString() const override {
+    std::stringstream ss("");
+    ss << " OrderID: " << orderID << " Product Name: " << name
+       << " Quantity: " << quantity;
+    return ss.str();
+  }
+
+  virtual size_t objectSize() const override {
+    auto objectSize = sizeof(Order);
+    return objectSize;
+  }
+
+  const std::string& getClassName() const override {
+    static std::string className = "com.example.Order";
+    return className;
+  }
+
+ private:
+  unsigned long orderID;
+  std::string name;
+  unsigned int quantity;
+
+  const std::string orderid_key = "orderid";
```

Comment:
Constant values should be ALL CAPS and static.

Reply:
is this to be consistent with a prior styling choice?  google style guide says to prefix consts (including static consts) with a 'k' followed by mixed case e.g. kOrderIdKey.

Reply:
so these are class members and we ran into a compiler error when making them static.

`/opt/local/geode-native/examples/cpp/customserializable/main.cpp:75:28: error: static data member of type 'const std::string' (aka 'const basic_string<char, char_traits<char>, allocator<char> >') must be initialized out of line

static const std::string ORDER_ID_KEY_ = "order_id";`

We did a search and found this, which makes it seem like without a separate hpp file that this would end up being awkward anyways:

https://stackoverflow.com/questions/11693690/const-array-declaration-in-c-header-file

Reply:
Yes... C++ is fun. You will need the header to declare your class and a source file to define it. This is all best practice. I should have caught that your class was defined in your main source file.